File: groupie/src/main/java/com/genius/groupie/ExpandableGroup.java
package com.genius.groupie;

import java.util.ArrayList;
import java.util.List;

/**
 * An ExpandableContentItem is one "base" content item with a list of children (any of which
 * may themselves be a group.)
 **/

public class ExpandableGroup extends NestedGroup {

    private boolean isExpanded = false;
    private final Group parent;
    private final List<Group> children = new ArrayList<>();

    public ExpandableGroup(Group expandableItem) {
        this.parent = expandableItem;
        ((ExpandableItem) expandableItem).setExpandableGroup(this);
    }

    @Override public void add(Group group) {
        super.add(group);
        if (isExpanded) {
            int itemCount = getItemCount();
            children.add(group);
            notifyItemRangeInserted(itemCount, group.getItemCount());
        } else {
            children.add(group);
        }
    }

    public boolean isExpanded() {
        return isExpanded;
    }

    public Group getGroup(int position) {
        if (position == 0) {
            return parent;
        } else {
            return children.get(position - 1);
        }
    }

    @Override public int getPosition(Group group) {
        if (group == parent) {
            return 0;
        } else {
            return 1 + children.indexOf(group);
        }
    }

    public int getGroupCount() {
        return 1 + (isExpanded ? children.size() : 0);
    }

    public void onToggleExpanded() {
        int oldSize = getItemCount();
        isExpanded = !isExpanded;
        int newSize = getItemCount();
        if (oldSize > newSize) {
            notifyItemRangeRemoved(newSize, oldSize - newSize);
        } else {
            notifyItemRangeInserted(oldSize, newSize - oldSize);
        }
    }
    
    private boolean dispatchChildChanges(Group group) {
        return isExpanded || group == parent;
    }

    @Override
    public void onChanged(Group group) {
        if (dispatchChildChanges(group)) {
            super.onChanged(group);
        }
    }

    @Override
    public void onItemInserted(Group group, int position) {
        if (dispatchChildChanges(group)) {
            super.onItemInserted(group, position);
        }
    }

    @Override
    public void onItemChanged(Group group, int position) {
        if (dispatchChildChanges(group)) {
            super.onItemChanged(group, position);
        }
    }

    @Override
    public void onItemChanged(Group group, int position, Object payload) {
        if (dispatchChildChanges(group)) {
            super.onItemChanged(group, position, payload);
        }
    }

    @Override
    public void onItemRemoved(Group group, int position) {
        if (dispatchChildChanges(group)) {
            super.onItemRemoved(group, position);
        }
    }

    @Override
    public void onItemRangeChanged(Group group, int positionStart, int itemCount) {
        if (dispatchChildChanges(group)) {
            super.onItemRangeChanged(group, positionStart, itemCount);
        }
    }

    @Override
    public void onItemRangeInserted(Group group, int positionStart, int itemCount) {
        if (dispatchChildChanges(group)) {
            super.onItemRangeInserted(group, positionStart, itemCount);
        }
    }

    @Override
    public void onItemRangeRemoved(Group group, int positionStart, int itemCount) {
        if (dispatchChildChanges(group)) {
            super.onItemRangeRemoved(group, positionStart, itemCount);
        }
    }

    @Override
    public void onItemMoved(Group group, int fromPosition, int toPosition) {
        if (dispatchChildChanges(group)) {
            super.onItemMoved(group, fromPosition, toPosition);
        }
    }
}

File: library-kotlin/src/main/java/com/wray/groupiekotlin/ExpandableGroup.kt
package com.wray.groupiekotlin

import android.util.Log

/**
 * The "collapsed"/header item of an expanded group.  Some part (or all) of it is a "toggle" to
 * expand the group.
 *
 * Collapsed:
 * - This
 *
 * Expanded:
 * - This
 * - Child
 * - Child
 * - etc
 *
 */
interface ExpandableItem: Group {
    fun setExpandableGroup(expandableGroup: ExpandableGroup)
}

/**
 * An `ExpandableGroup` is one "base" content item with a list of children (any of which
 * may themselves be a group.)
 **/

open class ExpandableGroup(
    private val parent: ExpandableItem,
    isExpanded: Boolean = false
): NestedGroup() {

    var isExpanded: Boolean = isExpanded
        set(value) {
            if (field != value) {
                onToggleExpanded()
            }
            field = value
        }

    private val children = mutableListOf<Group>()

    init {
        parent.setExpandableGroup(this)
    }

    override fun add(group: Group) {
        super.add(group)
        if (isExpanded) {
            val itemCount = itemCount
            children.add(group)
            notifyItemRangeInserted(itemCount, group.itemCount)
        }
        children.add(group)
    }

    override fun add(position: Int, group: Group) {
        super.add(position, group)
        children.add(position, group)
        if (isExpanded) {
            val notifyPosition = 1 + children.subList(0, position).itemCount
            notifyItemRangeInserted(notifyPosition, group.itemCount)
        }
    }

    override fun <T : Group> addAll(groups: Collection<T>) {
        if (groups.isEmpty()) return

        super.addAll(groups)

        if (isExpanded) {
            val itemCount = itemCount
            children.addAll(groups)
            notifyItemRangeInserted(itemCount, groups.itemCount)
        } else {
            this.children.addAll(groups)
        }
    }

    override fun <T : Group> addAll(position: Int, groups: Collection<T>) {
        if (groups.isEmpty()) {
            return
        }
        super.addAll(position, groups)
        this.children.addAll(position, groups)

        if (isExpanded) {
            val notifyPosition = 1 + children.subList(0, position).itemCount
            notifyItemRangeInserted(notifyPosition, groups.itemCount)
        }
    }

    override fun remove(group: Group) {
        if (!children.contains(group)) {
            Log.w(ExpandableGroup::class.java.simpleName, "Attempting to remove child from ExpandableGroup which does not contain it.")
            return
        }
        super.remove(group)

        if (isExpanded) {
            val position = getItemCountBeforeGroup(group)
            children.remove(group)
            notifyItemRangeRemoved(position, group.itemCount)
        } else {
            children.remove(group)
        }
    }

    override fun <T : Group> removeAll(groups: Collection<T>) {
        if (groups.isEmpty() || !this.children.containsAll(groups)) {
            Log.w(ExpandableGroup::class.java.simpleName, "Attempting to remove children from ExpandableGroup which does not contain them.")
            return
        }
        super.removeAll(groups)

        if (isExpanded) {
            this.children.removeAll(groups)
            groups.forEach { group ->
                val position = getItemCountBeforeGroup(group)
                children.remove(group)
                notifyItemRangeRemoved(position, group.itemCount)
            }
        } else {
            this.children.removeAll(groups)
        }
    }

    val childCount: Int
        get() = children.size

    fun onToggleExpanded() {
        val oldSize = itemCount
        isExpanded = !isExpanded
        val newSize = itemCount
        if (oldSize > newSize) {
            notifyItemRangeRemoved(newSize, oldSize - newSize)
        } else {
            notifyItemRangeInserted(oldSize, newSize - oldSize)
        }
    }

    private val Group.canDispatchChanges: Boolean
        get() = isExpanded || this === parent

    //region: NestedGroup methods

    override fun getGroup(position: Int): Group =
        if (position == 0) {
            parent
        } else {
            children[position - 1]
        }

    override val groupCount: Int
        get() = 1 + if (isExpanded) children.size else 0

    override fun getPosition(group: Group): Int {
        if (group === parent) {
            return 0
        }
        val index = children.indexOf(group)
        return if (index >= 0) {
            index + 1
        } else -1
    }

    override fun onChanged(group: Group) {
        if (group.canDispatchChanges) {
            super.onChanged(group)
        }
    }

    override fun onItemInserted(group: Group, position: Int) {
        if (group.canDispatchChanges) {
            super.onItemInserted(group, position)
        }
    }

    override fun onItemChanged(group: Group, position: Int) {
        if (group.canDispatchChanges) {
            super.onItemChanged(group, position)
        }
    }

    override fun onItemChanged(group: Group, position: Int, payload: Any) {
        if (group.canDispatchChanges) {
            super.onItemChanged(group, position, payload)
        }
    }

    override fun onItemRemoved(group: Group, position: Int) {
        if (group.canDispatchChanges) {
            super.onItemRemoved(group, position)
        }
    }

    override fun onItemRangeChanged(group: Group, positionStart: Int, itemCount: Int) {
        if (group.canDispatchChanges) {
            super.onItemRangeChanged(group, positionStart, itemCount)
        }
    }

    override fun onItemRangeChanged(group: Group, positionStart: Int, itemCount: Int, payload: Any) {
        if (group.canDispatchChanges) {
            super.onItemRangeChanged(group, positionStart, itemCount, payload)
        }
    }

    override fun onItemRangeInserted(group: Group, positionStart: Int, itemCount: Int) {
        if (group.canDispatchChanges) {
            super.onItemRangeInserted(group, positionStart, itemCount)
        }
    }

    override fun onItemRangeRemoved(group: Group, positionStart: Int, itemCount: Int) {
        if (group.canDispatchChanges) {
            super.onItemRangeRemoved(group, positionStart, itemCount)
        }
    }

    override fun onItemMoved(group: Group, fromPosition: Int, toPosition: Int) {
        if (group.canDispatchChanges) {
            super.onItemMoved(group, fromPosition, toPosition)
        }
    }

    //endregion: NestedGroup methods
}